### PR TITLE
Add sharing of requests

### PR DESCRIPTION
**Pull request by loldot**
- [x] New feature
- [ ] Bug fix
- [x] High impact

**Description of work:**
Allow sharing requests with users without normal privileges in line org/project org.
https://docs.fusion-dev.net/pr-194/domain/services/resources/tech-specs/ote-second-opinion/

**Testing:**
- [x] Can be tested
- [x] Automatic tests created / updated
- [x] Local tests are passing

**Checklist:**
- [x] Considered automated tests
- [x] Considered updating specification / documentation
- [x] Considered work items 
- [x] Considered security
- [x] Performed developer testing
- [x] Checklist finalized / ready for review

### CODE REVIEW

File: src/backend/api/Fusion.Resources.Domain/Commands/Requests/Sharing/ShareRequest.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                }
+
+                var existingRequests = await db.SharedRequests
+                    .Where(x => x.RequestId == request.RequestId && x.Source == request.Source)
+                    .ToDictionaryAsync(x => x.SharedWithId, cancellationToken);
+
+                foreach (var sharedWith in request.SharedWith)
+                {
+                    var person = await profileService.EnsurePersonAsync(sharedWith);
+                    if (existingRequests.TryGetValue(person!.Id, out var existingRequest))
```
Comment (xssic) [inline src/backend/api/Fusion.Resources.Domain/Commands/Requests/Sharing/ShareRequest.cs:55]: What if person not found? Just ignore ?

File: src/backend/api/Fusion.Resources.Domain/Services/ProfileServices.cs [hunk 2]
```diff
@@ -266,6 +266,23 @@ public async Task<DbExternalPersonnelPerson> EnsureExternalPersonnelAsync(string
...
+        public Task<IEnumerable<ResolvedPersonProfile>?> ResolveProfilesAsync(IEnumerable<PersonId> personIds)
+        {
+            return ResolveProfilesAsync(personIds.Select(x => (PersonIdentifier)x));
+        }
+
+        public async Task<IEnumerable<ResolvedPersonProfile>?> ResolveProfilesAsync(IEnumerable<PersonIdentifier> personIds)
+        {
+            try
+            {
+                return await profileResolver.ResolvePersonsAsync(personIds);
```
Comment (xssic) [inline src/backend/api/Fusion.Resources.Domain/Services/ProfileServices.cs:278]: Not sure if relevant here, but recently observed there is a limit of 500 items in list here.

File: src/backend/api/Fusion.Resources.Domain/Commands/Requests/Sharing/ShareRequest.cs (hunk 1 above)
Comment (loldot) [reply src/backend/api/Fusion.Resources.Domain/Commands/Requests/Sharing/ShareRequest.cs:55]: The validator for the request checks that persons exist. If person is null here, I think it's ok to throw a NRE.

File: src/backend/api/Fusion.Resources.Api/Controllers/Requests/ShareRequestsController.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+            var requests = await DispatchAsync(
+                new GetResourceAllocationRequests().ForAll().SharedWith(azureId)
+            );
+
+            return Ok(new ApiCollection<ApiResourceAllocationRequest>(requests.Select(x => new ApiResourceAllocationRequest(x))));
+        }
+
+        [HttpDelete("/resources/requests/internal/{requestId}/share/{sharedToAzureId}")]
+        public async Task<IActionResult> DeleteSharedRequests(Guid requestId, string sharedToAzureId)
```
[the remark this answers is not in the document]
Comment (loldot) [reply src/backend/api/Fusion.Resources.Api/Controllers/Requests/ShareRequestsController.cs:94]: good catch

**Review by xssic — APPROVED** [review]

**Pull request merged**